# Домашнее задание к лекции "Базовые понятия статистики"

## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [245]:
import pandas as pd

In [246]:
HorseData = pd.read_csv('https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv', na_values = "?", header = None)


In [247]:
HorseDataStudy = HorseData[{0,1,2,3,4,5,6,7}]
HorseDataStudy = HorseDataStudy.rename(columns = {0:'Surgery',1:'Age',2:'Hospital number',3:'rectal temperature',4:'pulse',5:'respiratory rate',6:'temperature of extremities',7:'peripheral pulse'})


In [248]:
HorseDataStudy.head()

Surgery  Age  Hospital number  rectal temperature  pulse  respiratory rate  \
0      2.0    1           530101                38.5   66.0              28.0   
1      1.0    1           534817                39.2   88.0              20.0   
2      2.0    1           530334                38.3   40.0              24.0   
3      1.0    9          5290409                39.1  164.0              84.0   
4      2.0    1           530255                37.3  104.0              35.0   

   temperature of extremities  peripheral pulse  
0                         3.0               3.0  
1                         NaN               NaN  
2                         1.0               1.0  
3                         4.0               1.0  
4                         NaN               NaN

### Surgery (категориальная величина)

In [249]:
HorseDataStudy['Surgery'].describe()

count    299.000000
mean       1.397993
std        0.490305
min        1.000000
25%        1.000000
50%        1.000000
75%        2.000000
max        2.000000
Name: Surgery, dtype: float64

In [250]:
print("Уникальные значения:")
HorseDataStudy['Surgery'].unique()

Уникальные значения:


array([ 2.,  1., nan])

In [251]:
print("Мода: ", HorseDataStudy['Surgery'].mode(dropna=False)[0])

Мода:  1.0


Выводы: Уникальных значений - 2, а так же один пропуск. Мода = 1 (то есть, 1 = Yes, it had surgery)


### Age  (категориальная величина)

In [252]:
HorseDataStudy['Age'].describe()

count    300.000000
mean       1.640000
std        2.173972
min        1.000000
25%        1.000000
50%        1.000000
75%        1.000000
max        9.000000
Name: Age, dtype: float64

In [253]:
print("Уникальные значения:")
HorseDataStudy['Age'].unique()

Уникальные значения:


array([1, 9])

In [254]:
HorseDataStudy['Age'].value_counts()

1    276
9     24
Name: Age, dtype: int64

In [255]:
print("Мода: ", HorseDataStudy['Age'].mode(dropna=False)[0])

Мода:  1


Выводы: Уникальных значений - 2. Исходя из набора уникальных значений и описания столбца, можно сделать вывод, что в некоторых записях допущена ошибка - указано значение 9, хотя такого значения нет в описании. Мода = 1 (то есть, 1 = Adult horse)

### Hospital number  (дискретная величина)

In [256]:
HorseDataStudy['Hospital number'].describe()

count    3.000000e+02
mean     1.085889e+06
std      1.529801e+06
min      5.184760e+05
25%      5.289040e+05
50%      5.303055e+05
75%      5.347275e+05
max      5.305629e+06
Name: Hospital number, dtype: float64

In [257]:
print("Всего записей:" ,len(HorseDataStudy['Hospital number']))
print("Уникальных значений:" ,len(HorseDataStudy['Hospital number'].unique()))

Всего записей: 300
Уникальных значений: 284


In [258]:
print("Мода: ")
print( HorseDataStudy['Hospital number'].mode(dropna=False))

Мода: 
0      527544
1      527916
2      528151
3      528469
4      528729
5      528890
6      528904
7      528931
8      528996
9      529424
10     529461
11     529796
12     530526
13     530693
14     532349
15    5279822
dtype: int64


Выводы: Уникальных значений - 284. Записи о некоторых лошадях занесены дважды.Всего таких 16. 

### rectal temperature (непрерывная величина)

In [259]:
HorseDataStudy['rectal temperature'].describe()

count    240.000000
mean      38.167917
std        0.732289
min       35.400000
25%       37.800000
50%       38.200000
75%       38.500000
max       40.800000
Name: rectal temperature, dtype: float64

Выводы: Всего 240 записей, то есть 60 пропусков.

### pulse (непрерывная величина)

In [260]:
HorseDataStudy['pulse'].describe()

count    276.000000
mean      71.913043
std       28.630557
min       30.000000
25%       48.000000
50%       64.000000
75%       88.000000
max      184.000000
Name: pulse, dtype: float64

Выводы: Всего 276 записей, то есть 24 пропуска.

### respiratory rate (непрерывная величина)

In [261]:
HorseDataStudy['respiratory rate'].describe()

count    242.000000
mean      30.417355
std       17.642231
min        8.000000
25%       18.500000
50%       24.500000
75%       36.000000
max       96.000000
Name: respiratory rate, dtype: float64

Выводы: Всего 242 записей, то есть 58 пропусков.

### temperature of extremities (категориальная величина)

In [262]:
HorseDataStudy['temperature of extremities'].describe()

count    244.000000
mean       2.348361
std        1.045054
min        1.000000
25%        1.000000
50%        3.000000
75%        3.000000
max        4.000000
Name: temperature of extremities, dtype: float64

In [263]:
print("Мода: ")
print( HorseDataStudy['temperature of extremities'].mode(dropna=False))

Мода: 
0    3.0
dtype: float64


Выводы: Всего 244 записей, то есть 56 пропусков. Мода = 3, то есть значение 'Cool' встречается чаще всего.

### peripheral pulse  (категориальная величина)

In [264]:
HorseDataStudy['peripheral pulse'].describe()

count    231.000000
mean       2.017316
std        1.042428
min        1.000000
25%        1.000000
50%        2.000000
75%        3.000000
max        4.000000
Name: peripheral pulse, dtype: float64

In [265]:
print("Мода: ")
print( HorseDataStudy['peripheral pulse'].mode(dropna=False))

Мода: 
0    1.0
dtype: float64


Выводы: Всего 231 записей, то есть 69 пропусков. Мода = 1, то есть значение 'Normal' встречается чаще всего.

### Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

- An elevated temp may occur due to infection.
- temperature may be reduced when the animal is in late shock
- normal temp is 37.8
- pulse: reflection of the heart condition: 30 -40 is normal for adults
- respiratory rate: normal rate is 8 to 10
- temperature of extremities:
  a subjective indication of peripheral circulation
  possible values: 1 = Normal 2 = Warm 3 = Cool 4 = Cold
- peripheral pulse
  possible values are: 1 = normal 2 = increased 3 = reduced 4 = absent

### rectal temperature:

In [266]:
HorseDataStudyCopy = HorseDataStudy.copy()
HorseDataStudyCopy = HorseDataStudyCopy.dropna(subset = {'rectal temperature'})

In [267]:
q1 = HorseDataStudyCopy['rectal temperature'].quantile(0.25)
q3 = HorseDataStudyCopy['rectal temperature'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = HorseDataStudyCopy[HorseDataStudyCopy['rectal temperature'].between(lower_bound, upper_bound, inclusive=True)]
pd.concat([HorseDataStudyCopy, remove_outliers]).drop_duplicates(keep=False)


Surgery  Age  Hospital number  rectal temperature  pulse  \
20       1.0    1           530157                39.9   72.0   
44       1.0    1           535407                35.4  140.0   
75       1.0    9           534092                39.7  100.0   
80       1.0    1           527518                36.4   98.0   
99       2.0    1           530002                39.6  108.0   
118      1.0    1           533983                36.5   78.0   
141      2.0    1           522979                36.0   42.0   
238      2.0    1           528702                36.1   88.0   
251      2.0    1           527940                36.6   42.0   
259      1.0    1           527734                40.8   72.0   
281      2.0    1           530624                40.0   78.0   
298      1.0    1           530612                36.5  100.0   

     respiratory rate  temperature of extremities  peripheral pulse  
20               60.0                         1.0               1.0  
44               24.0                         3.0               3.0  
75                NaN                         3.0               3.0  
80               35.0                         3.0               3.0  
99               51.0                         3.0               3.0  
118              30.0                         1.0               NaN  
141              30.0                         NaN               NaN  
238               NaN                         3.0               3.0  
251              18.0                         3.0               3.0  
259              42.0                         3.0               3.0  
281               NaN                         3.0               3.0  
298              24.0                         3.0               3.0

Значения температуры выходят за пределы нормы. Но возможно, соответствуют реальным значениям. Предлагаю оставить эти строчки и значения для дальнейшего анализа.

### pulse:

In [268]:
HorseDataStudyCopy = HorseDataStudy.copy()
HorseDataStudyCopy = HorseDataStudyCopy.dropna(subset = {'pulse'})

In [269]:
q1 = HorseDataStudyCopy['pulse'].quantile(0.25)
q3 = HorseDataStudyCopy['pulse'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = HorseDataStudyCopy[HorseDataStudyCopy['pulse'].between(lower_bound, upper_bound, inclusive=True)]
pd.concat([HorseDataStudyCopy, remove_outliers]).drop_duplicates(keep=False)


Surgery  Age  Hospital number  rectal temperature  pulse  \
3        1.0    9          5290409                39.1  164.0   
41       2.0    9          5288249                39.0  150.0   
55       1.0    9          5282839                38.6  160.0   
255      1.0    9          5294539                38.8  184.0   
275      1.0    9          5297159                38.8  150.0   

     respiratory rate  temperature of extremities  peripheral pulse  
3                84.0                         4.0               1.0  
41               72.0                         NaN               NaN  
55               20.0                         3.0               NaN  
255              84.0                         1.0               NaN  
275              50.0                         1.0               3.0

In [270]:
remove_outliers['pulse'].describe()

count    271.000000
mean      70.258303
std       26.082484
min       30.000000
25%       48.000000
50%       64.000000
75%       88.000000
max      146.000000
Name: pulse, dtype: float64

Значения пульса очень большие в данных выбросах. И это сопровождается повышенной температурой. А так же значением Age = 9, которое, возможно, является ошибкой, так как среди возможных значений этого столбца есть только 1 и 2.
Предлагается игнорировать (удалить) строчки с такими выбросами.

### respiratory rate:

In [271]:
HorseDataStudyCopy = HorseDataStudy.copy()
HorseDataStudyCopy = HorseDataStudyCopy.dropna(subset = {'respiratory rate'})

In [272]:
q1 = HorseDataStudyCopy['respiratory rate'].quantile(0.25)
q3 = HorseDataStudyCopy['respiratory rate'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = HorseDataStudyCopy[HorseDataStudyCopy['respiratory rate'].between(lower_bound, upper_bound, inclusive=True)]
pd.concat([HorseDataStudyCopy, remove_outliers]).drop_duplicates(keep=False)

Surgery  Age  Hospital number  rectal temperature  pulse  \
3        1.0    9          5290409                39.1  164.0   
39       1.0    9          5277409                39.2  146.0   
41       2.0    9          5288249                39.0  150.0   
82       1.0    9          5290759                38.1  100.0   
84       1.0    1           529849                37.8   60.0   
103      1.0    9          5292489                38.0  140.0   
106      1.0    1           528668                38.3   52.0   
120      1.0    1           528682                39.4   54.0   
125      1.0    1           528461                38.0   42.0   
186      1.0    1          5278331                39.3   64.0   
208      1.0    1           535338                37.8   88.0   
229      1.0    9           534597                38.5  120.0   
244      1.0    9          5291409                38.2  124.0   
255      1.0    9          5294539                38.8  184.0   
265      2.0    1           528590                 NaN   56.0   
269      1.0    1           527883                39.3   64.0   
295      1.0    1           533886                 NaN  120.0   

     respiratory rate  temperature of extremities  peripheral pulse  
3                84.0                         4.0               1.0  
39               96.0                         NaN               NaN  
41               72.0                         NaN               NaN  
82               80.0                         3.0               1.0  
84               80.0                         1.0               3.0  
103              68.0                         1.0               1.0  
106              96.0                         NaN               3.0  
120              66.0                         1.0               1.0  
125              68.0                         4.0               1.0  
186              90.0                         2.0               3.0  
208              80.0                         3.0               3.0  
229              70.0                         NaN               NaN  
244              88.0                         1.0               3.0  
255              84.0                         1.0               NaN  
265              68.0                         3.0               1.0  
269              90.0                         2.0               3.0  
295              70.0                         4.0               NaN

Значение выбросов 'respiratory rate' выглядят очень большими по сравнению с нормой (8-10). Кроме того, в описании датасета указано, что в измерениях этой величины есть большие "флуктуации", и её "полезность" для анализа вызывает сомнения. Возможно, стоит ничего не делать с этими выбросами.

### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [273]:
HorseDataStudy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
Surgery                       299 non-null float64
Age                           300 non-null int64
Hospital number               300 non-null int64
rectal temperature            240 non-null float64
pulse                         276 non-null float64
respiratory rate              242 non-null float64
temperature of extremities    244 non-null float64
peripheral pulse              231 non-null float64
dtypes: float64(6), int64(2)
memory usage: 18.8 KB


In [274]:
HorseDataStudy.dropna(thresh=6).info() #будем считать, что нужно, чтобы было заполнено минимум 6 столбцов

<class 'pandas.core.frame.DataFrame'>
Int64Index: 276 entries, 0 to 299
Data columns (total 8 columns):
Surgery                       275 non-null float64
Age                           276 non-null int64
Hospital number               276 non-null int64
rectal temperature            238 non-null float64
pulse                         267 non-null float64
respiratory rate              238 non-null float64
temperature of extremities    236 non-null float64
peripheral pulse              224 non-null float64
dtypes: float64(6), int64(2)
memory usage: 19.4 KB


In [275]:
HorseDataStudy = HorseDataStudy.dropna(thresh=4)

#### Surgery

In [276]:
# Surgery
HorseDataStudy[HorseDataStudy['Surgery'].isna()]

Surgery  Age  Hospital number  rectal temperature  pulse  \
132      NaN    1           534572                38.0   48.0   

     respiratory rate  temperature of extremities  peripheral pulse  
132              20.0                         3.0               1.0

 - Предлагаю удалить эту запись, так как она всего одна, а так же потому, что сделать предположение о том, какое значение пропущено, кажется крайне затруднительным.

In [277]:
HorseDataStudy = HorseDataStudy.dropna(subset = {'Surgery'})

#### rectal temperature

Для 'rectal temperature' предлагаю заменить пропущенные значения средним, так как стандартное отклонение у этой величины небольшое. А пропущенных записей довольно много.


In [278]:
HorseDataStudy['rectal temperature'].fillna(HorseDataStudy['rectal temperature'].mean(), inplace=True)

#### pulse

In [279]:
HorseDataStudy[HorseDataStudy['pulse'].isna()]

Surgery  Age  Hospital number  rectal temperature  pulse  \
5        2.0    1           528355           38.168619    NaN   
52       2.0    1           529483           38.168619    NaN   
58       1.0    1           528298           38.168619    NaN   
78       1.0    1           530693           38.168619    NaN   
83       1.0    1          5279822           38.000000    NaN   
93       2.0    1           530310           38.168619    NaN   
115      2.0    1           533723           38.168619    NaN   
117      2.0    1          5290482           39.500000    NaN   
126      1.0    1           530384           38.700000    NaN   
136      1.0    1           530693           38.168619    NaN   
151      1.0    1           529399           39.300000    NaN   
159      2.0    1           528134           38.168619    NaN   
174      2.0    1           518476           38.168619    NaN   
182      1.0    1          5279822           38.000000    NaN   
204      1.0    1           529528           39.200000    NaN   

     respiratory rate  temperature of extremities  peripheral pulse  
5                 NaN                         2.0               1.0  
52                NaN                         1.0               1.0  
58               20.0                         4.0               3.0  
78                NaN                         3.0               3.0  
83               24.0                         3.0               3.0  
93                NaN                         3.0               3.0  
115              40.0                         3.0               1.0  
117               NaN                         3.0               3.0  
126               NaN                         3.0               1.0  
136               NaN                         3.0               3.0  
151               NaN                         4.0               3.0  
159              12.0                         1.0               1.0  
174               NaN                         3.0               4.0  
182              24.0                         3.0               3.0  
204              23.0                         3.0               1.0

In [280]:
HorseDataStudy['pulse'].describe()

count    275.000000
mean      72.000000
std       28.646219
min       30.000000
25%       48.000000
50%       64.000000
75%       88.000000
max      184.000000
Name: pulse, dtype: float64

 Удалять польностью пропуски для pulse кажется неверным, так как их много. Предлагается заменить их медианным значением, так как оно характеризует выборку.

In [281]:
HorseDataStudy['pulse'].fillna(HorseDataStudy['pulse'].median(), inplace=True)
HorseDataStudy['pulse'].fillna(HorseDataStudy['pulse'].median()).isna().sum()

0

#### respiratory rate 

In [282]:
HorseDataStudy[HorseDataStudy['respiratory rate'].isna()]

Surgery  Age  Hospital number  rectal temperature  pulse  \
5        2.0    1           528355           38.168619   64.0   
7        1.0    1           529607           38.168619   60.0   
9        2.0    9          5299629           38.300000   90.0   
29       2.0    1           535415           37.700000   48.0   
32       1.0    1           529427           37.200000   60.0   
37       1.0    1           533847           37.800000   72.0   
38       2.0    1           528996           38.600000   52.0   
40       1.0    1           529498           38.168619   88.0   
43       1.0    1           534069           38.168619  120.0   
45       2.0    1           529827           38.168619  120.0   
50       2.0    1           529766           38.600000   46.0   
52       2.0    1           529483           38.168619   64.0   
59       2.0    1           528904           38.168619   96.0   
61       2.0    1           527957           38.500000   60.0   
66       2.0    1           534784           37.900000   72.0   
72       1.0    1          5299253           37.700000   56.0   
75       1.0    9           534092           39.700000  100.0   
78       1.0    1           530693           38.168619   64.0   
81       1.0    1           534756           37.300000   40.0   
93       2.0    1           530310           38.168619   64.0   
112      1.0    1           534963           38.168619   40.0   
117      2.0    1          5290482           39.500000   64.0   
126      1.0    1           530384           38.700000   64.0   
129      1.0    1           527698           38.400000   49.0   
134      1.0    1           527524           38.168619  100.0   
136      1.0    1           530693           38.168619   64.0   
137      1.0    1           534324           38.000000   48.0   
138      2.0    1           532349           38.000000   56.0   
142      1.0    1           530402           38.168619  120.0   
146      1.0    1           534115           38.200000   48.0   
148      1.0    1           529272           38.300000  132.0   
151      1.0    1           529399           39.300000   64.0   
152      1.0    1           533836           38.168619  100.0   
164      1.0    1           528523           38.700000   76.0   
173      1.0    1           529729           38.168619  104.0   
174      2.0    1           518476           38.168619   64.0   
179      2.0    1           528151           38.500000  100.0   
188      1.0    1           533887           38.500000   60.0   
201      1.0    1           528641           38.500000   86.0   
205      2.0    1           528151           38.500000  100.0   
238      2.0    1           528702           36.100000   88.0   
245      2.0    1           528904           38.168619   96.0   
249      2.0    1           528169           38.200000   48.0   
254      1.0    1           533738           38.168619   80.0   
256      1.0    1           533871           37.500000   72.0   
271      1.0    1           534626           37.700000   80.0   
274      1.0    1           534624           38.168619   76.0   
281      2.0    1           530624           40.000000   78.0   
284      2.0    1           530439           38.500000   54.0   

     respiratory rate  temperature of extremities  peripheral pulse  
5                 NaN                         2.0               1.0  
7                 NaN                         3.0               NaN  
9                 NaN                         1.0               NaN  
29                NaN                         2.0               1.0  
32                NaN                         2.0               1.0  
37                NaN                         NaN               3.0  
38                NaN                         1.0               1.0  
40                NaN                         3.0               3.0  
43                NaN                         3.0               4.0  
45                NaN                         4.0     

In [283]:
HorseDataStudy['respiratory rate'].describe()

count    241.000000
mean      30.460581
std       17.666102
min        8.000000
25%       18.000000
50%       25.000000
75%       36.000000
max       96.000000
Name: respiratory rate, dtype: float64

 Удалять польностью пропуски для respiratory rate кажется неверным, так как их много. Предлагается заменить их медианным значением, так как оно характеризует выборку.

In [284]:
HorseDataStudy['respiratory rate'].fillna(HorseDataStudy['respiratory rate'].median(), inplace=True)

#### temperature of extremities и peripheral pulse

Предлагаю заменить пропуски на значения моды, так как значения катеогориальные, и удалять совсем их нет смысла

In [287]:
HorseDataStudy['temperature of extremities'].fillna(HorseDataStudy['temperature of extremities'].mode()[0], inplace=True)


In [288]:
HorseDataStudy['peripheral pulse'].fillna(HorseDataStudy['peripheral pulse'].mode()[0], inplace=True)


In [295]:
# Сформирован датафрейм без пропусков:

In [296]:
HorseDataStudy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 290 entries, 0 to 299
Data columns (total 8 columns):
Surgery                       290 non-null float64
Age                           290 non-null int64
Hospital number               290 non-null int64
rectal temperature            290 non-null float64
pulse                         290 non-null float64
respiratory rate              290 non-null float64
temperature of extremities    290 non-null float64
peripheral pulse              290 non-null float64
dtypes: float64(6), int64(2)
memory usage: 20.4 KB


In [297]:
HorseDataStudy.describe()

Surgery         Age  Hospital number  rectal temperature  \
count  290.000000  290.000000     2.900000e+02          290.000000   
mean     1.406897    1.606897     1.055851e+06           38.168619   
std      0.492104    2.121880     1.493197e+06            0.665862   
min      1.000000    1.000000     5.184760e+05           35.400000   
25%      1.000000    1.000000     5.289108e+05           37.825000   
50%      1.000000    1.000000     5.303145e+05           38.168619   
75%      2.000000    1.000000     5.347002e+05           38.500000   
max      2.000000    9.000000     5.305629e+06           40.800000   

            pulse  respiratory rate  temperature of extremities  \
count  290.000000        290.000000                  290.000000   
mean    71.586207         29.537931                    2.451724   
std     27.949308         16.228908                    0.987507   
min     30.000000          8.000000                    1.000000   
25%     48.000000         20.000000                    1.000000   
50%     64.000000         25.000000                    3.000000   
75%     88.000000         35.000000                    3.000000   
max    184.000000         96.000000                    4.000000   

       peripheral pulse  
count        290.000000  
mean           1.810345  
std            1.016420  
min            1.000000  
25%            1.000000  
50%            1.000000  
75%            3.000000  
max            4.000000

## Дополнительная часть (необязательная)

Выполнить задания 1-3 для всего набора данных.

#### ПРИМЕЧАНИЕ
Домашнее задание сдается ссылкой на репозиторий [GitHub](https://github.com/).
Не сможем проверить или помочь, если вы пришлете:
- файлы;
- архивы;
- скриншоты кода.

Все обсуждения и консультации по выполнению домашнего задания ведутся только на соответствующем канале в slack.

##### Как правильно задавать вопросы аспирантам, преподавателям и коллегам
Прежде чем задать вопрос, попробуйте найти ответ в интернете. Навык самостоятельного поиска информации — один из важнейших. Каждый практикующий специалист любого уровня делает это ежедневно.

Сформулируйте вопрос по алгоритму:  
1) Что я делаю?  
2) Какого результата я ожидаю?  
3) Как фактический результат отличается от ожидаемого?  
4) Что я уже попробовал сделать, чтобы исправить проблему?  

По возможности прикрепите к вопросу скриншоты либо ссылки на код. Не выкладывайте все решение, оставляйте только проблемный и воспроизводимый участок кода.